### DISCUSSION TITLE
Is the creation of an iModel possible using only open-source/free software/services?

### DISCUSSION THREAD

**Discussion by donvi-bz** [iTwin/itwinjs-core#1659] [Q&A] [answered]
From what I can tell, this library, while open source, requires paying for at least one product or service from Bentley. It seems like we either need one of the paid modeling software (like OpenPlant or MicroStation), or a subscription to ProjectWise. Could someone tell me if this is correct? Also, if this is, could we just get the ProjectWise subscription for one person and leave it at that? Lastly, in case this type of question shouldn't be asked here, where should it be asked?

**Comment by kabentley** [comment, marked as answer]
Yes, this is the best place to ask this question.

You can create, query, visualize, and analyze iModels (among many other things) using only the free and open source iTwin.js library, without any paid licenses or subscriptions. You can use any modeling software to create your BIM data, and use the free [iTwin Synchronizer](https://www.bentley.com/en/products/product-line/digital-twins/itwin-synchronizer) tool to convert to an iModel. That will create a static [snapshot iModel](https://www.itwinjs.org/learning/backend/accessingimodels/#snapshot-imodels). If you re-run the synchronizer, it will create a new snapshot iModel. 

However, if you wish to track changes with changesets stored in the cloud, you do need to have a subscription for Bentley's iTwin service (and users will each need a login, etc.) Think of it as GitHub for iModels. There are many other services available with an iTwin subscription (issue tracking, web visualization/review, clash detection, design insights, etc.) As explained [here](https://github.com/imodeljs/imodeljs/blob/master/core/backend/src/imodeljs-native-LICENSE.md) that's our commercial motivation for creating the library.

In version 3.0 it will be possible to create interactive desktop (Electron) or web applications that directly create/edit iModels. Again, change tracking will require an iTwin subscription. But if you don't want to track changes, everything else works without any license/subscription.

HTH,
Keith

**Reply by donvi-bz** [reply]
Thank you for this reply, I guess somewhere along researching I missed the point of the syncronizer. To make it worse, I even have it installed on my computer. I will mention though, websites like [this](https://www.bentley.com/en/i-models/creating-i-models/creating-i-models) and [this](https://communities.bentley.com/products/digital-twin-cloud-services/itwin-services/w/imodelhub/43161/create-an-imodel) make it seem like there is only one way to make the iModel, and its possible to go through lots of documentation of one of those links without finding the other, or finding much information on the syncronizer. 

This is the fourth time I've been active on this github, and everyone here is great, but some of this documentation really gets confusing. Again, thank you for this, I think it is the last piece of information I needed clarified.

**Reply by kabentley** [reply]
@donvi-bz one clarification -  i mentioned iTwin Synchronizer above, and i meant [iTwin Snapshot](https://www.itwinjs.org/learning/tutorials/itwin-snapshot-app/). As the names imply, one creates snapshots and one synchronizes with the hub. Frankly I don't understand why there couldn't be one download that could be configured either way (both are free and probably 95% identical). Sorry about that.

All of this does take a bit of explaining, since really there's nothing like iModels or iTwins anywhere else in the world. I will see about adding more clarifications to the links above.

**Reply by Moult** [reply]
Is there any free software stack / library to create an iModel? Is iModel open data? Where can I find its specifications and schema?

Is there any free software backend available for iTwinJS? It seems the critical piece @bentley/imodeljs-native is closed, akin to a [proprietary driver blob](https://en.wikipedia.org/wiki/Proprietary_device_driver#Problems). Given such a critical component, are you sure you can compare it to "Think of it as GitHub for iModels"? In contrast, Git, the core in Github, is truly free.

Also, the [following paragraph](https://www.itwinjs.org/learning/backend/accessingimodels/#snapshot-imodels) suggests that we are not allowed to create our own backend for iTwin unless we are [licensed by Bentley](https://github.com/iTwin/itwinjs-core/blob/master/core/backend/src/imodeljs-native-LICENSE.md). Therefore, if we want to use free software, we are limited to static snapshot iModels and can never benefit from the dynamic nature of a changing BIM model.

> @bentley/imodeljs-native implements the authentication and access enforcement expressed by iModel owners for (non-Snapshot) iModels. User authentication and access rights are a service of Bentley Systems, as a part of if its iTwin Services offerings. Bentley may also license third parties to supply similar services. This is Bentley Systems' commercial motivation for creating iTwin.js. Any attempt to circumvent or disrupt this checking is a violation of the license agreement.

Is this interpretation correct? If so, isn't this crippleware?

**Comment by kabentley** [comment]
The page you reference and my reply above have links to the free software to create iModels, and there are apis in iTwin.js for creating your own connector (and it's an area of very active development now.) I don't know how to respond to the "is iModel Open" question, since I've already directly answered it in my reply above and in the documentation. [Here](https://www.itwinjs.org/bis/) is the link to the schema documentation. 

As i explain above, you are free to use the library to create, visualize, analyze, and even edit iModels. I don't know how you conclude that you can't create your own backend, but that's not true. What you have to pay for is our cloud service to track changes, just like GitHub (which is not free for commercial use). Seems like a pretty logical comparison to me.

In my not-so-humble and very biased opinion, iTwin.js is the most comprehensive, most performant, most advanced technology for infrastructure digital twins anywhere. The vast majority of it is here, open sourced and developed in the open under the MIT license. If your interpretation of our commercial incentive for granting you access to that huge and valuable resource is that it is "crippleware" (whatever that means), feel free to move on and ignore us.

**Reply by Moult** [reply]
@kabentley thanks for the link to the schema documentation. I did not realise BIS was the iModel schema. I could not however find whether or not the schema was owned by Bentley, or if it was open source, or whether there was an external body driving the development of the schema, or if it was just Bentley. Can you help clarify this? My concern is what happened with DXF - where development and specifications were driven and published by Autodesk alone, and as a result DXF became a second class citizen. There are potential parallels between the DWG/DXF industry problem and iModel+BentleySubscription vs Snapshot iModel. To help put these concerns at rest, can you confirm that development of iModel is independent of Bentley and ideally vendor-neutral?

I was trying to get started with development and both the [this page](https://www.bentley.com/en/i-models/creating-i-models/creating-i-models) only lists Bentley products on how to create an iModel, or requires me to be part of the Bentley Developer Network, and the [getting started tutorial](https://developer.bentley.com/create-imodel/) required a Bentley account to progress past the first step. I followed your link to [iTwin Snapshot](https://www.itwinjs.org/learning/tutorials/itwin-snapshot-app/) but was unable to find a Linux download. Can you please help point me to the source code for iTwin Snapshot?

Also, can you confirm that it is legally allowed for someone to create a backend for iModels that fully replicates the services of iTwinJS? (Or even better, that a fully free backend already exists?) This is what I meant by "crippleware" - my impression after reading the sites was that right now, it is only possible to practically benefit from iModels with a Bentley subscription, or, I need explicit permission and be licensed by Bentley to build my own. Here are the quotes that led me to that conclusion:

> Snapshot iModels **should not be used in any production workflows** as they do not provide security, change history, and other benefits of an **iTwin Subscription.**

> The @bentley/imodeljs-native module is written in C++, is delivered as a platform-specific binary, and is **not open source**. It contains the SQLite code that directly accesses iModel files and implements all the low-level APIs upon which @bentley/imodeljs-backend is based. It is **always required for all backends**.

> @bentley/imodeljs-native implements the authentication and access enforcement expressed by iModel owners for (non-Snapshot) iModels. User authentication and access rights are a service of Bentley Systems, as a part of if its iTwin Services offerings. **Bentley may also license third parties to supply similar services**. This is Bentley Systems' commercial motivation for creating iTwin.js. Any attempt to circumvent or disrupt this checking is a violation of the license agreement.

> They cannot be modified and therefore have no timeline. The ability to synchronize non-Snapshot iModels and track and store changes to iModels is a service offered by Bentley **(or third parties licensed by Bentley)** for a fee. Synchronized iModels can be offline temporarily when a network connection is not available.

> Note that only the imodeljs-backend module has a dependency on imodeljs-native. **Frontend applications do not require a right-to-run** (the iTwin.js libraries are MIT Licensed), though **they may connect to a backend that does**.


... and so on.

I hope I am wrong and that a 100% free software ecosystem can be built around iModel that is also vendor neutral. However the documentation right now sends a mixed message. Hopefully, it is simply a wording problem which can be easily solved.

**Reply by kabentley** [reply]
> I could not however find whether or not the schema was owned by Bentley, or if it was open source, or whether there was an external body driving the development of the schema, or if it was just Bentley. Can you help clarify this? 

The schema is an open source as part of iTwin.js You should note that the reason it is called the *Base* Infrastructure Schema is it defines the classes used, in code, in iTwin.js. The expectation is that other domain-specific schemas will be developed by Bentley and others (including proprietary end-user schemas), which may or may not be contributed back to iTwin.js. We hope that there will be many, but to date most of the actual work has been performed by Bentley (in several cases in close cooperation with Siemens, for example.) There is no external body driving the schemas, because Bentley has been the only contributor. We would welcome working with others willing to contribute (for example with Building Smart, and others). Keep in mind that schemas often require code to enforce/interpret/integrate their contents, so contributing will involve implementations of the classes defined in the schemas. That's more work than most domain experts who aren't programmers are willing to do. 

> Also, can you confirm that it is legally allowed for someone to create a backend for iModels.

Yes, i can confirm, and many already are. Though when you say "fully replicates the services of iTwin.js", i presume you mean "fully replicates Bentley's subscription services implemented with iTwin.js". As i mentioned before, the only service the license prohibits is the one that deals with managing changesets, timelines, and briefcases. I agree the language above doesn't make that clear enough. Some of that is due to terminology evolution. I will see if we can make that more clear. 

> I hope I am wrong and that a 100% free software ecosystem can be built around iModel that is also vendor neutral. 

Our commitment is that we will make our library, and the data created with it, open and useable without the need for a subscription from Bentley. We and others are free to use that library to make products that *do* require a license/subscription, and we hope many will. So when you say "100% free software ecosystem", I presume you mean free speech, not free beer.

**Reply by Moult** [reply]
Cheers, this makes sense and is a fantastic sign of Bentley driving positive change in the industry.

> So when you say "100% free software ecosystem", I presume you mean free speech, not free beer.

Yes, indeed :) Devs need to money to eat! Just one more point of clarification:

> As i mentioned before, the only service the license prohibits is the one that deals with managing changesets, timelines, and briefcases.

So, if I want to build a service using all of these MIT libraries, and if I rebuild the black box @bentley/imodeljs-native as fully free software, then am I allowed to use my own free software version of @bentley/imodeljs-native to build a service that manages changesets, timelines, and briefcases, and offer it as an alternative clone to Bentley's services, or is there some legal text prohibiting that?

If my interpretation is correct, and so long as we build our own @freesoftware/imodeljs-native-free-edition, we have the freedom to fully clone Bentley's services, then this is truly fantastic news. I guess the only observation others from the free software community might have when they stumble across this issue is that as you've mentioned, there isn't currently much of a community behind iTwinJS - it's basically just Bentley. As you say you would welcome working with others willing to contribute, would you be interested in turning iTwinJS into a more community driven development? I think the OSArch community would love to support this work, especially if it can mean a fully free digital twin stack. For example, personally I'd love to see how systems like IFC and Brickschema can integrate.

**Reply by kabentley** [reply]
> So, if I want to build a service using all of these MIT libraries, and if I rebuild the black box @bentley/imodeljs-native as fully free software, then am I...

Ugh. There'd be nothing to prohibit it, but it would be a terrible outcome from my perspective. First of all the native-api isn't documented very well, and is all marked as @internal so we can change it without breaking abi. Then, it's all platform specific and would be a lot of work to duplicate, much less maintain over time, particularly on all platforms (iOS/Android are a mess). So, the outcome of that would be that people would "hang back" and not stay current with the platform's evolution. That's already a "problem". I think I'd rather just remove any restrictions in the license and hope people won't feel the need to duplicate it. Frankly, I think the current license was just fear-of-the-dark at the beginning of the project, and I really don't think the existence of alternative services from others competing with Bentley is actually bad for us, net. 

The main reason @bentley/imodeljs-native isn't open source *isn't* that we are trying to keep it private, it's that it would be a lot of work. Building it currently involves internal Bentley build tools/pipelines, Have you ever had the misfortune of building Electron/Node from scratch? It's at least that complicated.

**Reply by Moult** [reply]
@kabentley thanks for the informative reply. I understand your concerns, and definitely there is a balance between the highly technical developer world (i.e. building electron from scratch) and user accessibility. I also think that opening up is a net benefit for Bentley - users are upskilling and getting more technical, and perhaps the pipelines can be made a bit more accessible. With it closed, indeed I predict a similar issue with the proprietary driver blobs - where the open ones tend to lag behind - so iTwin gets perceived as legally a majority open source, but not really "open" culturally. It would also perhaps be a necessary step towards opening to the community and not just stay as a Bentley thing - if that's something you're keen on.

Would you be interested in collaborating with the OSArch community and building some integrations into existing free software in the AEC space? That might be a fantastic way to engage the existing FOSS community!

**Reply by Moult** [reply]
A bit of food for thought - I know we're talking MIT here and not GPL, but GPL has some interesting tidbits about how being open is not just about providing the source, but also about the practicality of external contributors to build, setup, install, etc - so perhaps opening up would be a great way to address what sounds like a rather tricky build process. See section 4.2.2 here: https://softwarefreedom.org/resources/2008/compliance-guide.html

**Reply by kabentley** [reply]
@Moult Thank you very much for your frank insights. To be clear, I think that if our project is perceived as "legally but not culturally open" we'll end up in the worst of all worlds - having spent a lot of effort/$ to create an open ecosystem that isn't used where it could be beneficial. Nobody wants that, most of all me.

> "Would you be interested in collaborating with the OSArch community.."

absolutely. How do we get that started?

**Reply by Moult** [reply]
Perhaps the best way to start is to perhaps present iTwinJS in one of the OSArch monthly meetup sessions where we do show and tell and tech previews (i.e. to a technical audience). It might also be good to reach out on the forums or look at some of the existing free software initiatives (IFC.JS, Topologic, OpenProject, FreeCAD, etc) and see if there is a potential integration that can be built - after all, we are all hackers here and the best way to collaborate is to write code together! If you want to be less formal and say hello as individuals, the forums and live chat are a great place. @DuncanLithgow is pretty awesome at helping coordinate the OSArch community, so perhaps something can be organised together?

**Reply by Moult** [reply]
A bump to see if any progress has been made on this. Is @bentley/imodeljs-native still closed? I didn't notice any outreach to the open source CAx/BIM community?

**Reply by kabentley** [reply]
We have been working on this project internally and should be ready in a few months. As I mentioned, it was a non-trivial project. We have had to separate some internal dependencies and some housekeeping to eliminate some non-essential cruft the node addon doesn't need. One project has been to create an interface for our BRrep layers to eliminate our direct dependency on Parasolid, so that interface can be supplied by other libraries (I hereby invite @mgooding to connect with you on that project).

I know you mentioned GPL above, but I don't think that's for us. We're currently planning on releasing under Apache 2.0, and switching all the @itwin projects from MIT to Apache2. Do you have an opinion on license form? 

I thought that someone here was going to reach out to you guys. @calebmshafer ?

**Reply by mgooding** [reply]
@Moult - expect an email from me regarding the BRep project shortly. While we've been working on the abstracted Parasolid interface for our library, we're still gathering feedback from community members on what a realistic open-source implementation would look like.

**Comment by ColinKerr** [comment]
@Moult @donvi-bz we just made our repository for all BIS schemas public: [bis-schemas](https://github.com/iTwin/bis-schemas). 

This repository contains work in progress and released versions of every schema Bentley has created for iModels including schemas for commercial Bentley applications that are not open sourced.  All schemas (and WIP versions) are published to npmjs.org in a form that makes them easy to consume in with iTwin.js e.g. [Bis Core](https://www.npmjs.com/package/@bentley/bis-core-schema) or [Structural Analysis](https://www.npmjs.com/package/@bentley/structural-analysis-schema)

It is the repository we use internally for schema development and we have open sourced the tools we use to validate schemas for release*.  So it is possible to develop BIS schemas using the same workflow we use internally.

*Some of our tooling has a dependency on the core code contained in the imodeljs-native library.  This means the process is not open all the way down but we have made everything else public.

Making the repository public means we are making the commitment that BIS schemas for our commercial products will be made available with an open license.

**Comment by DuncanLithgow** [comment]
Really interesting project. Great that the BIS-schemas are now public. Is there a press release for that I can tweet from the OSArchOrg twitter account?

Would be so great if Bentley finds a good way to fit into the opensource / free software. 

@kabentley one way to get started is to mention @OSArchOrg ( https://twitter.com/OSArchOrg ) whenever it makes sense to push a message to the opensource community in AEC. I'll keep an eye on what turns up. I've pushed my first cautious retweet. https://twitter.com/OSArchOrg/status/1462525882275450895

This thread isn't really the place for this collaboration - if that's what we're starting here. We could meet on our forum community.osarch.org (lots of FOSS eyeballs) or somewhere else. Please let us know what makes sense for you.
